src/ziggurat/mapper.clj

```diff
@@ -15,47 +15,71 @@
   (fn [message]
     (let [topic-entity-name          (name topic-entity)
           new-relic-transaction-name (str topic-entity-name ".handler-fn")
-          metric-namespace           (str topic-entity-name ".message-processing")]
+          default-namespace          "message-processing"
+          metric-namespaces          [topic-entity-name default-namespace]
+          additional-tags            {:topic-name topic-entity-name}
+          default-namespaces         [default-namespace]
+          success-metric             "success"
+          retry-metric               "retry"
+          skip-metric                "skip"
+          failure-metric             "failure"]
       (nr/with-tracing "job" new-relic-transaction-name
         (try
           (let [start-time  (.toEpochMilli (Instant/now))
                 return-code (mapper-fn message)
                 end-time    (.toEpochMilli (Instant/now))]
-            (metrics/report-time (str topic-entity-name ".handler-fn-execution-time") (- end-time start-time))
+            (metrics/report-time [topic-entity-name "handler-fn-execution-time"] (- end-time start-time))
             (case return-code
-              :success (metrics/increment-count metric-namespace "success")
-              :retry (do (metrics/increment-count metric-namespace "retry")
-                         (producer/retry message topic-entity))
-              :skip (metrics/increment-count metric-namespace "skip")
-              :block 'TODO
+              :success (do (metrics/increment-count metric-namespaces success-metric additional-tags)
```

Comment:
just a suggestion, since you are calling the same function with different namespaces here for all cases, why not pull it out into a function? IMO that would remove a lot of duplication

Reply:
done